syntax = "proto3";

// Registry Service
service RegistryService {
  rpc register(RegisterRequest) returns (RegisterResponse);
  rpc deregister(DeregisterRequest) returns (DeregisterResponse);
  rpc populate_finger_table(PopulateFingerTableRequest) returns (stream NodeInfoItem);
}

// Registry Service for client - implemented to split server from client
service RegistryClientService {
  rpc get_chord_info(Empty) returns (stream NodeInfoItem);
  rpc connect(Empty) returns (ConnectResponse);
}

// Node Service
service NodeService {
  rpc get_finger_table(Empty) returns (stream NodeInfoItem);
  rpc save(SaveRequest) returns (NodeActionResponse);
  rpc remove(FindRemoveRequest) returns (NodeActionResponse);
  rpc find(FindRemoveRequest) returns (NodeActionResponse);
  rpc connect(Empty) returns (ConnectResponse);
  rpc request_key_value(KeyValueRequest) returns (stream KeyValueResponse);
}

// Messages that used several times
message Empty {}

message ConnectResponse {
  string type = 1;
}

// Messages for RegistryService
message RegisterRequest {
  string ipaddr = 1;
  int32 port = 2;
}

message RegisterResponse {
  int64 id = 1;
  string message = 2;
}

message DeregisterRequest {
  int64 id = 1;
}

message DeregisterResponse {
  bool status = 1;
  string message = 2;
}

message PopulateFingerTableRequest {
  int64 id = 1;
}

message NodeInfoItem {
  int64 id = 1;
  string address = 2;
}

// Messages for NodeService
message SaveRequest {
  string key = 1;
  string text = 2;
}

message FindRemoveRequest {
  string key = 1;
}

message NodeActionResponse {
  bool status = 1;
  string message = 2;
}

message KeyValueRequest {
  int64 id = 1;
}

message KeyValueResponse {
  string key = 1;
  string text = 2;
}
